WebApp/recommendations.py

```diff
@@ -0,0 +1,164 @@
+        book['rank'] = i + 1
+
+        # for some reason, some of the text fields have newlines appended to them
+        book['title'] = book['title'].strip()
+        book['author'] = book['author'].strip()
+        top_books.append(book)
+
+    return top_books
+
+
+def map_user(q, V):
+    # map new user to concept space by q*V
+    user_to_concept = np.matmul(q, V)
+    # map user back to itme space with user_to_concept * VT
+    result = np.matmul(user_to_concept, V.T)
+    return result
+
+def map_user_to_features(p, features):
+    p_sparse = scipy.sparse.csr_matrix(p)
+    # map new user to concept space by p*features
+    user_to_concept = p_sparse.dot(features)
+    # map user back to itme space with user_to_concept * featuresT
+    result = user_to_concept.dot(features.T).todense()
+    return result.T
+
+def get_predictions(p, q, user_bias, item_bias, global_bias):
+    pred_ratings = np.zeros(len(q))
+    for i in range(len(q)):
+        pred = global_bias + user_bias + item_bias[i] + np.dot(p, q[i])
+        pred_ratings[i] = pred
+    return pred_ratings
+
+def partial_fit(new_user_ratings, q, item_bias, global_bias):
+
+    # create array of indices of books this user has actually rated
+    indices = []
+    for i in range(len(new_user_ratings)):
+        if new_user_ratings[i] != 0:
+            indices.append(i)
+
+    # Hyperparams
+    learning_rate = 0.07
+    user_bias_reg = 0.001
+    P_reg = 0.001
+
+    # 5 updates per rated book
+    iterations = 5
+
+    n_factors = q.shape[1]
+    cols = q.shape[0]
+    print("n_factors: {}".format(n_factors))
+
+    # 1. set the user_bias for this user
+    new_user_bias = 0
+
+    # 2. set up new random P
+    mu, sigma = 0, 0.1
+    p = np.random.normal(mu, (sigma / n_factors), n_factors)
+
+    # 3. computer small number of iterations of SGD
+    for iteration in range(iterations):
+        for i in indices:
+            rating = new_user_ratings[i]
+            pred = global_bias + new_user_bias + item_bias[i] + np.dot(p, q[i])
+            error = rating - pred
+
+            # update P
+            for f in range(n_factors):
+                p_update = learning_rate * (error * q[i][f] - P_reg * p[f])
+                p[f] += p_update
+
+            # update user bias
+            ub_update = learning_rate * (error - user_bias_reg * new_user_bias)
+            new_user_bias += ub_update
+
+    return get_predictions(p, q, new_user_bias, item_bias, global_bias)
+
+def log_rank(predictions_partial_fit, predictions_features, user_ratings, books, weight_feature, num_books):
+
+    # create tuple of book_id and rating for each method, then sort
+    partial_fit_ratings = []
+    feature_ratings = []
+    for i in range(len(books)):
+        partial_fit_ratings.append((i, predictions_partial_fit[i]))
+        feature_ratings.append((i, predictions_features[i]))
+    
+    partial_fit_ratings = sorted(partial_fit_ratings, key=lambda x: x[1], reverse=True)
```

Comment:
These can all be wayy simplified using numpy, but I don't mind as long as it works. I can touch up on them later.

Reply:
yeah feel free to revamp log rank lol, its dirty right now